File: backends/spring-java/src/main/java/dev/stackverse/backend/message/LanguageResolver.java
package dev.stackverse.backend.message;

import java.util.List;
import java.util.Locale;
import java.util.Set;
import org.springframework.stereotype.Component;

@Component
public class LanguageResolver {
    public static final String DEFAULT_LANGUAGE = "en";

    private final MessageRepository messageRepository;

    public LanguageResolver(MessageRepository messageRepository) {
        this.messageRepository = messageRepository;
    }

    public String resolve(String lang, String acceptLanguage) {
        Set<String> supported = messageRepository.findDistinctLanguages();
        if (lang != null && supported.contains(lang)) {
            return lang;
        }
        if (acceptLanguage != null && !acceptLanguage.isBlank()) {
            List<Locale.LanguageRange> ranges;
            try {
                ranges = Locale.LanguageRange.parse(acceptLanguage);
            } catch (IllegalArgumentException exception) {
                ranges = List.of();
            }
            for (Locale.LanguageRange range : ranges) {
                String code = Locale.forLanguageTag(range.getRange()).getLanguage();
                if (supported.contains(code)) {
                    return code;
                }
            }
        }
        return DEFAULT_LANGUAGE;
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/message/LanguageResolver.kt
package dev.stackverse.backend.message

import org.springframework.stereotype.Component
import java.util.Locale

const val DEFAULT_LANGUAGE = "en"

/**
 * SPEC rule 8: explicit `lang` parameter → first supported language in `Accept-Language`
 * (quality-ordered) → `en`. Unsupported values fall back down the chain, never error.
 * "Supported" means at least one message exists in that language.
 */
@Component
class LanguageResolver(private val messageRepository: MessageRepository) {

    fun resolve(lang: String?, acceptLanguage: String?): String {
        val supported = messageRepository.findDistinctLanguages()
        if (lang != null && lang in supported) {
            return lang
        }
        if (!acceptLanguage.isNullOrBlank()) {
            val ranges = runCatching { Locale.LanguageRange.parse(acceptLanguage) }.getOrDefault(emptyList())
            for (range in ranges) {
                val code = Locale.forLanguageTag(range.range).language
                if (code in supported) {
                    return code
                }
            }
        }
        return DEFAULT_LANGUAGE
    }
}
